File: src/p1.js
import * as R from "rambda";

//create a range

const range = R.range(1, 1000);

//check if number divisible by 3
const divBy3 = (n) => n % 3 === 0;

//check if number is divisible by 5
const divBy5 = (n) => n % 5 === 0;

//compose a function that will check if a number is divisible by 3 or 5
const divBy3or5 = R.either(divBy3, divBy5);

//filter on range, and curry on the inputrange
const result = R.pipe(R.filter(divBy3or5), R.sum)(range);

console.log(result);

File: src/p1.ml
let divBy3 n = 
  if n mod 3 = 0 then true else false 

let divBy5 n = 
  if n mod 5 = 0 then true else false 

let divBy3Or5 n = 
  divBy3 n || divBy5 n 

(* ref : https://batsov.com/articles/2022/10/31/ocaml-tips-implementing-a-range-function/*)
let rec range start stop =
  List.init (stop - start) ( fun i -> i + start)

let rec filter f l =
  match l with
  | [] -> []
  | h :: t -> if f h then h :: filter f t else filter f t
  
let rec sum l =
  match l with
  | [] -> 0
  | h :: t -> h + sum t
  
let result =
  range 1 1000
  |> filter divBy3Or5
  |> sum
  
let () = 
  Printf.printf "%d\n" result